src/environmentd/src/http.rs

```diff
@@ -271,65 +275,95 @@ impl AuthedClient {
         &mut self,
         request: HttpSqlRequest,
     ) -> Result<SqlHttpExecuteResponse, anyhow::Error> {
+        // This API prohibits executing statements with responses whose
+        // semantics are at odds with an HTTP response.
+        fn check_prohibited_stmts(stmt: &Statement<Raw>) -> Result<(), anyhow::Error> {
+            let execute_responses = Plan::generated_from(stmt.into())
+                .into_iter()
+                .map(ExecuteResponse::generated_from)
+                .flatten()
+                .collect::<Vec<_>>();
+
+            if execute_responses.iter().any(|execute_response| {
+                matches!(
+                    execute_response,
+                    ExecuteResponseKind::Fetch
+                        | ExecuteResponseKind::SetVariable
+                        | ExecuteResponseKind::Tailing
+                        | ExecuteResponseKind::CopyTo
+                        | ExecuteResponseKind::CopyFrom
+                        | ExecuteResponseKind::Raise
+                        | ExecuteResponseKind::DeclaredCursor
+                        | ExecuteResponseKind::ClosedCursor
+                )
+            }) && !matches!(
+                stmt,
+                // Both `SelectStatement` and `CopyStatement` generate
+                // `PeekPlan`, but `SELECT` should be permitted and `COPY` not.
+                Statement::Select(mz_sql::ast::SelectStatement { query: _, as_of: _ })
+            ) {
+                anyhow::bail!("unsupported via this API: {}", stmt.to_ast_string());
+            }
+
+            Ok(())
+        }
+
+        let mut stmt_groups = vec![];
```

Comment:
My understanding of what `stmt_groups` is doing is that it sets `num_stmts` which is used by `start_transaction` to decide on the type of transaction started. The pgwire extended protocol doesn't know how many statements it'll get, which is why it sets that to 1 (==Started) and then has the option to eagerly commit transactions if it wants. Because we do know here exactly how many statements are present over the extended protocol, I think this can safely not have these groups, but just a `Vec<(Statement, Params)>` and use the total statement count for `start_transaction`.

Reply:
Removing `stmt_groups` prevents multiple DDL statements from getting used in the same extended request, which PG supports https://github.com/MaterializeInc/materialize/pull/14896

I'm glad to remove it, but want to make sure that the impact is acknowledged.

Reply:
Wait what? Why do the statement groups allow this to happen? Oh...it's because the txn is started with a count > 1 blergh. Might need to think about this.

Reply:
From discussion over slack: we should keep the statement groups because it keeps the same facade up as the pgwire extended protocol (#14038).